### PR TITLE
implement logExecutionTimeBpp

### PR DESCRIPTION
**Pull request by ajax-malov-n**
+ Implementation of BeanPostProcessor

### CODE REVIEW

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        val clazz: Class<*> = bean.javaClass
+        try {
+            val superMethod = clazz.getDeclaredMethod(method.name, *(method.parameterTypes ?: emptyArray()))
+            if (superMethod.isAnnotationPresent(LogExecutionTime::class.java)) {
+                return true
+            }
+        } catch (_: NoSuchMethodException) {
+        }
+        return false
+    }
```
Comment (viacheslav-torbin) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: IMHO, try-catch looks ugly here. You can probably do it like this:
```kotlin
        val matchingMethod = clazz.declaredMethods.firstOrNull { 
            it.name == method.name && it.parameterTypes.contentEquals(method.parameterTypes)
        }
        return matchingMethod?.isAnnotationPresent(LogExecutionTime::class.java) ?: false
```

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        if (bean.javaClass.methods.any { it.isAnnotationPresent(LogExecutionTime::class.java) }) {
+            beanMap[beanName] = bean
+        }
+        return bean
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = beanMap[beanName]
+        if (beanClass != null) {
```
Comment (ajax-butriy-v) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: May be **return beanClass?.let { logic } ?: bean** (although in fact it has same logic as yours)

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        return beanMap[beanName]?.let { beanClass ->
+            Proxy.newProxyInstance(
+                beanClass.javaClass.getClassLoader(),
+                getAllInterfaces(beanClass.javaClass).toTypedArray()
+            )
+            { _, method, args ->
+                if (isMethodAnnotatedInSuperclass(method, beanClass)) {
+                    val before = System.nanoTime()
+                    val result = method.invoke(bean, *(args ?: emptyArray()))
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: what if the function throw exception?

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                    log.info("Execution time of ${method.name} is ${(System.nanoTime() - before) / 1_000_000} ms")
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: Never inline parameters into log message.

❌ Incorrect usage:
```kotlin
val some = 123
logger.info("Logging parameter=$some using inlining into the message")
```

✅ correct:
```kotlin
val some = 123
logger.info("Logging parameter={} using message pattern", some)

```

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        } ?: bean
+    }
+
+    private fun getAllInterfaces(clazz: Class<*>): Set<Class<*>> {
+        val interfaces = mutableSetOf<Class<*>>()
+        var currentClass: Class<*>? = clazz
+        while (currentClass != null) {
+            interfaces.addAll(currentClass.interfaces)
+            currentClass = currentClass.superclass
+        }
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt:78]: let's practice kotlin's `sequence`.

Please, rework this function by using [generate-sequence](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/generate-sequence.html)

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        return interfaces
+    }
+
+    private fun isMethodAnnotatedInSuperclass(method: Method, bean: Any): Boolean {
+        val clazz: Class<*> = bean.javaClass
+        try {
+            val superMethod = clazz.getDeclaredMethod(method.name, *(method.parameterTypes ?: emptyArray()))
+            return superMethod.isAnnotationPresent(LogExecutionTime::class.java)
+        } catch (_: NoSuchMethodException) {
+            return false
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: let's rework this function as well.

In general it's bad practice to use exceptions for a flow control.
It's better to refactor in the next way:
1. obtain all methods
2. find the required method in list

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]
Let's add extra functionality to make your BPP less trivial.
Please, add option to enable/disable profiling using config properties

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.stereotype.Component
+import java.lang.reflect.Method
+import java.lang.reflect.Proxy
+
+
+@ConditionalOnProperty(name = ["logExecutionTime.enabled"], havingValue = "true")
+@Component
+class LogExecutionTimeAnnotationBeanPostProcessor : BeanPostProcessor {
+
+    private val beanMap = HashMap<String, Pair<Class<*>, List<Method>>>()
```
Comment (ajax-butriy-v) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: may be extract `Pair<Class<*>, List<Method>>` to a separate data class for readability?

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+    private fun isRequiredProxy(
+        methods: List<Method>,
+        proxiedMethod: Method,
+    ) = methods.any { originalMethod ->
+        areMethodsEqual(
+            originalMethod,
+            proxiedMethod
+        )
+    }
```
Comment (ajax-butriy-v) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: all other members use `return` statement, while this one has expression body

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTime.kt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package systems.ajax.malov.stockanalyzer.config.beanpostprocessor
+
+@Retention(AnnotationRetention.RUNTIME)
+@Target(AnnotationTarget.FUNCTION)
+annotation class LogExecutionTime()
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTime.kt]:
```suggestion
annotation class LogExecutionTime
```

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    private fun getAllInterfaces(clazz: Class<*>): Array<Class<*>> {
+        return generateSequence(clazz) { it.superclass }
+            .flatMap { it.interfaces.asSequence() }
+            .toSet()
+            .toTypedArray()
+    }
+
+    private fun areMethodsEqual(originalMethod: Method, proxiedMethod: Method): Boolean {
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: Let's practice infix function?

```kotlin
// invocation part:
val equals = proxiedMethod isEqualTo originalMethod
```
Isn't it beautiful?

Or, at least extension function.

The suggestion in **optional**, so, if you don't want apply the suggestion then just resolve the thread.

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        proxiedMethod: Method,
+        bean: Any,
+        args: Array<out Any>?,
+    ): Any? {
+        val before = System.nanoTime()
+        return try {
+            proxiedMethod.invoke(bean, *(args ?: emptyArray<Any?>()))
+        } finally {
+            val executionTimeInMs = (System.nanoTime() - before) / 1_000_000
+            log.info("Execution time of {} is {} ms", proxiedMethod.name, executionTimeInMs)
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: `System.nanoTime()` is used but logged as `ms`.

Also, the variable `executionTimeInMs` lies about it's measurement units.

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    ): Boolean {
+        return methods.any { originalMethod ->
+            areMethodsEqual(
+                originalMethod,
+                proxiedMethod
+            )
+        }
+    }
+
+    private fun logMethodExecution(
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: the function does:
- invoke target bean function
- log execution time

But `logMethodExecution` tells only about the logging and nothing about the invocation

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+                }
+            }
+        } ?: bean
+    }
+
+    private fun isRequiredProxy(
+        methods: List<Method>,
+        proxiedMethod: Method,
+    ): Boolean {
+        return methods.any { originalMethod ->
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: What complexity we have here?

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+import java.lang.reflect.Proxy
+
+
+@ConditionalOnProperty(name = ["logExecutionTime.enabled"], havingValue = "true")
+@Component
+class LogExecutionTimeAnnotationBeanPostProcessor : BeanPostProcessor {
+
+    private val beanMap = HashMap<String, ClassWithAnnotatedMethods>()
+
+    data class ClassWithAnnotatedMethods(val clazz: Class<*>, val methods: List<Method>)
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: should it be private?

----

I'd like to see an internal classes after all members of the host class

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

**Review by ajax-gnatiuk-s — COMMENTED** [review]
LGTM

should we add some tests?

File: src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt (hunk 13 above)
Comment (ajax-gnatiuk-s) [reply src/main/kotlin/systems/ajax/malov/stockanalyzer/config/beanpostprocessor/LogExecutionTimeAnnotationBeanPostProcessor.kt]: we talk about this offline. Are going to make complexity closer to `O(1)`?

**Review by ajax-gnatiuk-s — APPROVED** [review]
LGTM

**Pull request merged**